app/vmctl/mimir/lazyreader.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	return &LazyBlockReader{
+		ID:          block.ID,
+		SegmentsNum: block.SegmentsNum,
+		fs:          fs,
+	}, nil
+}
+
+func (lbr *LazyBlockReader) initialize() error {
+	lbr.mx.Lock()
+	defer lbr.mx.Unlock()
+	if lbr.reader != nil {
+		return nil
+	}
+	// fetching block and parse it and store it in lbr.reader
+	temp, err := lbr.mkTempDir()
+	if err != nil {
+		return fmt.Errorf("failed to create temp dir: %s", err)
+	}
+
+	meta, err := lbr.fetchFile(metaFilename)
+	if err != nil {
+		return err
+	}
+	if err := lbr.writeFile(temp, metaFilename, meta); err != nil {
+		log.Printf("failed to write meta file: %s", err)
+		return err
+	}
+	idx, err := lbr.fetchFile(indexFilename)
+	if err != nil {
+		log.Printf("failed to fetch index file %q: %s", indexFilename, err)
+		return err
+	}
+	if err := lbr.writeFile(temp, indexFilename, idx); err != nil {
+		return err
+	}
+
+	for i := 1; i <= lbr.SegmentsNum; i++ {
+		// segments formats has format 1b06d
+		// https://github.com/grafana/mimir/blob/main/pkg/storage/tsdb/bucketindex/index.go#L32
+		chunkName := fmt.Sprintf("%06d", i)
+		blockChunkPath := filepath.Join("chunks", chunkName)
+		chunk, err := lbr.fetchFile(blockChunkPath)
+		if err != nil {
+			log.Printf("failed to fetch chunk file: %q: %s", chunkName, err)
+			return err
+		}
+		if err := lbr.writeFile(temp, blockChunkPath, chunk); err != nil {
+			log.Printf("failed to write chunk file: %q: %s", chunkName, err)
+			return err
+		}
+	}
+
+	// Set postingDecoder to nil because
+	// If it is nil then a default decoder is used, compatible with Prometheus v2.
+	pb, err := tsdb.OpenBlock(nil, temp, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to open block %q: %s", lbr.ID, err)
+	}
+	lbr.reader = pb
+	blockID := lbr.ID.String()
+	blockPath := filepath.Join(temp, blockID)
+	if err := os.RemoveAll(blockPath); err != nil {
+		log.Printf("failed to remove temp dir: %s", err)
+		return fmt.Errorf("failed to remove temp dir: %s", err)
+	}
+	return nil
+}
+
+// Index returns an IndexReader over the block's data.
+func (lbr *LazyBlockReader) Index() (tsdb.IndexReader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Index()
+}
+
+// Chunks returns a ChunkReader over the block's data.
+func (lbr *LazyBlockReader) Chunks() (tsdb.ChunkReader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Chunks()
+}
+
+// Tombstones returns a tombstones.Reader over the block's deleted data.
+func (lbr *LazyBlockReader) Tombstones() (tombstones.Reader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Tombstones()
+}
+
+// Meta provides meta information about the block reader.
+func (lbr *LazyBlockReader) Meta() tsdb.BlockMeta {
+	if err := lbr.initialize(); err != nil {
+		log.Printf("error get Block Meta: %s; return empty block", err)
+		return tsdb.BlockMeta{}
```

Comment:
It's better to return `err` explicitly or assign it to the `lbr.err` and check it at caller side.

Reply:
add method that will return an error and add the field to the LazyBlockReader struct. Check this error into caller side